router/module.go

```diff
@@ -225,117 +225,162 @@ func (am AppModule) OnChanCloseConfirm(ctx sdk.Context, portID, channelID string
-			denom = unprefixedDenom
-
-			// The denomination used to send the coins is either the native denom or the hash of the path
-			// if the denomination is not native.
-			denomTrace := transfertypes.ParseDenomTrace(unprefixedDenom)
-			if denomTrace.Path != "" {
-				denom = denomTrace.IBCDenom()
-			}
-		} else {
-			prefixedDenom := transfertypes.GetDenomPrefix(packet.GetDestPort(), packet.GetDestChannel()) + newData.Denom
-			denom = transfertypes.ParseDenomTrace(prefixedDenom).IBCDenom()
-		}
-		unit, err := sdk.ParseUint(newData.Amount)
-		if err != nil {
-			channeltypes.NewErrorAcknowledgement(err.Error())
-		}
-		var token = sdk.NewCoin(denom, sdk.NewIntFromUint64(unit.Uint64()))
 
-		var timeout time.Duration
-		if parsedReceiver.Timeout.Nanoseconds() > 0 {
-			timeout = parsedReceiver.Timeout
-		} else {
-			timeout = am.forwardTimeout
-		}
+	token := sdk.NewCoin(denomOnThisChain, amountInt)
 
-		var retries uint8
-		if parsedReceiver.ForwardRetries != nil {
-			retries = *parsedReceiver.ForwardRetries
-		} else {
-			retries = am.retriesOnTimeout
-		}
+	var timeout time.Duration
+	if metadata.Timeout.Nanoseconds() > 0 {
+		timeout = metadata.Timeout
+	} else {
+		timeout = am.forwardTimeout
+	}
 
-		err = am.keeper.ForwardTransferPacket(ctx, nil, packet, data.Sender, parsedReceiver, token, retries, timeout, []metrics.Label{})
-		if err != nil {
-			am.keeper.RefundForwardedPacket(ctx, packet, am.refundTimeout)
-			ack = channeltypes.NewErrorAcknowledgement(err.Error())
-		}
+	var retries uint8
+	if metadata.Retries != nil {
+		retries = *metadata.Retries
+	} else {
+		retries = am.retriesOnTimeout
+	}
+
+	err = am.keeper.ForwardTransferPacket(ctx, nil, packet, data.Sender, data.Receiver, metadata, token, retries, timeout, []metrics.Label{})
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(err.Error())
 	}
-	return ack
+
+	// returning nil ack will prevent WriteAcknowledgement from occurring for forwarded packet.
+	// This is intentional so that the acknowledgement will be written later based on the ack/timeout of the forwarded packet.
+	return nil
 }
 
 // OnAcknowledgementPacket implements the IBCModule interface
 func (am AppModule) OnAcknowledgementPacket(ctx sdk.Context, packet channeltypes.Packet, acknowledgement []byte, relayer sdk.AccAddress) error {
-	if err := am.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer); err != nil {
-		return err
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		am.keeper.Logger(ctx).Error("packetForwardMiddleware error parsing packet data from ack packet",
+			"sequence", packet.Sequence,
+			"src-channel", packet.SourceChannel, "src-port", packet.SourcePort,
+			"dst-channel", packet.DestinationChannel, "dst-port", packet.DestinationPort,
+			"error", err,
+		)
+		return am.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
```

Comment:
Why are you calling the underlying app here, instead of returning error

Reply:
I did this in order to avoid making assumptions on the expected packet data. In the case a middleware further down the stack is expecting a packet that is not `transfertypes.FungibleTokenPacketData`, this will allow it to fall through.